### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by timon**
This PR improves overall stability of the library:

- The base class is changed from `IOError` to `Exception`, as there is nothing wrong with input/output if the twinkly responds with unexpected `code` or `401 Not authorized`
- Removed passing of `**kwargs` to exception constructor, as Python exceptions [do not handle keyword args](https://github.com/python/typeshed/issues/2344#issuecomment-407151110)
- HTTP responses are checked for being successful prior JSON parsing is attempted on body
- 401 responses are translated into `TokenExpiredError`
- `X-Auth-Token` invalidation is handled by retrying the failed request once. 

My Twinkly remembers only the last token issued, so sending a command from mobile app invalidates the session of `xled`. The two last points work around the issue, retrying the failed request if the Twinkly responds with 401 error, and raising the `TokenExpiredError` if the second request fails as well.

### CODE REVIEW

File: xled/response.py [hunk 1]
```diff
@@ -59,6 +59,9 @@ def data(self):
                     "No response to create application response data from"
                 )
 
+            if not self.response.ok:
+                raise ApplicationError(self.response.text)
```
Comment (scrool) [inline xled/response.py]: I'm not sure if raising ApplicationError() here is a good idea. If one would explicitly run `raise_for_status(propagate=True)` (see bellow) they would get `HTTPError()` exception. But here they would get ApplicationError for de facto HTTP error.

I guess better option would be to call here self.response.raise_for_status() and let underlying requests to raise `HTTPError()`.

What do you think?

File: xled/exceptions.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ class AuthenticationError(XledException):
     """Authentication handshake wasn't successful"""
 
 
-class TokenExpiredError(XledException):
+class TokenExpiredError(ApplicationError):
```
Comment (scrool) [inline xled/exceptions.py]: Why this change?

File: xled/control.py [hunk 3]
```diff
@@ -72,6 +72,20 @@ def session(self):
             assert self._session
         return self._session
 
+    def session_get(self, url, retriable = True):
```
Comment (scrool) [inline xled/control.py]: I'm wondering if this doesn't duplicate `auto_refresh_token` flag of `BaseUrlChallengeResponseAuthSession`? Granted, it isn't set by default in the code.

Or did I misunderstood what you try to achieve here?

File: xled/control.py [hunk 4]
```diff
@@ -86,6 +86,17 @@ def session_get(self, url, retriable = True):
         else:
             return ApplicationResponse(response)
 
+    def session_post(self, url, **kwargs):
```
Comment (scrool) [inline xled/control.py]: See my comment for token expiry on get requests.

File: xled/control.py (hunk 3 above)
Comment (timon) [reply xled/control.py]: The `TokenExpiredError` is raised only when calculated token's time to live is expired. In my case, the token was invalidated by another client. Initially I found it while controlling the lights both from mobile app and from python REPL, but the simplest steps to reproduce is described in 6cccbb4e:

```python
ctrl1 = xled.control.ControlInterface(ip)
ctrl2 = xled.control.ControlInterface(ip)

ctrl1.is_on() # => True
ctrl2.is_on() # => True
ctrl1.is_on() # => raises an exception
```
on master, this snippet will throw a JSON decoding error, as 401 response does not have any JSON data in body.

6cccbb4e adds a check for HTTP status, fe8e337 converts the error to TokenExpiredError, and the last two commit handle the case by retrying the failed requests once.

File: xled/response.py (hunk 1 above)
Comment (timon) [reply xled/response.py]: `raise_for_status` will raise if the response does not have `code` field in response JSON.
Unfortunately, the endpoint for querying configured timers does not include this field:
[1](https://xled-docs.readthedocs.io/en/latest/rest_api.html#get-timer), [2](https://github.com/timon/twinkly/blob/master/API.md#get-configured-timers)

HTTPError is a better choice, agreed.

File: xled/control.py (hunk 3 above)
Comment (scrool) [reply xled/control.py]: I see. I like to handle this kind of issues. On the other hand I'd rather like to have this handling on lower level - by `auth.BaseUrlChallengeResponseAuthSession`. From a quick look it might be possible to hook to responses similarly how `ChallengeResponseAuth` does that. I'm not sure if all the retry handling would fit nicely though.

What do you think?

File: xled/exceptions.py (hunk 2 above)
Comment (scrool) [reply xled/exceptions.py]: The semantic of `ApplicationError()` is to catch issues received from payload found in HTTP body. Technically it would be application that runs on the device. `XledException()` are exceptions used to handle various issues within the `xled` package. Until now `TokenExpiredError()` wasn't raised on error caught from the device.

If there is a need to handle HTTP error 401 Forbidden I'd prefer to use separate exception that would be based on ApplicationError as you proposed here.

**Comment by timon** [comment]
Rebased the branch onto current `develop`.

File: xled/control.py (hunk 3 above)
Comment (timon) [reply xled/control.py]: If I understood you, #36 should do the right thing.

**Comment by timon** [comment]
Closing as #36 achieves the same goal with less changes
